API-3292 Add option for horizontal rendering to Formlet.Ocelot.Checkbox.checkboxSet

citizengabe opened the pull request:
## What does this pull request do?

We want to be able to render a checkbox set horizontally. In order to do so, we need to be able to specify the number of columns to be used during rendering as we do for tabular select.

## Where should the reviewer start?
All at once.

## How should this be manually tested?

Automated testing should be sufficient. It would be nice to add a storybook for these components though.

## Other Notes:

Since `Formlet.Ocelot.Checkbox.Render` implements `Semigroup`, `columns` must be a `Monoid`. While we could select any arbitrary monoid (such as `First Int`), we decide to use `Array Int` to preserve values and ordering. While it's unlikely any `render` function will make use of this sequence, we shouldn't make assumptions.

Also, I'm requesting a review from Wenbo since he's most familiar with `tabularSelect` upon which this solution is based.

citizengabe committed:
Add an optional columns property to checkboxSet

We want to be able to render a checkbox set horizontally. In order to do
so, we need to be able to specify the number of columns to be used
during rendering as we do for tabular select.

Since `Formlet.Ocelot.Checkbox.Render` implements `Semigroup`, `columns`
must be a `Monoid`. While we could select any arbitrary monoid (such as
`First Int`), we decide to use `Array Int` to preserve values and
ordering. While it's unlikely any `render` function will make use of
this sequence, we shouldn't make assumptions.

citizengabe committed:
Render horizontally when columns is specified

boygao1992 commented on src/Formlet/Ocelot/Checkbox.purs:
```diff
@@ -13,16 +13,20 @@ import Data.Set as Data.Set
-        , readonly :: Boolean
-        }
-    )
+    { columns :: Array Int -- NOTE positive integer (> 0), we use Array for its Semigroup instance and to collect column values
```
I didn't take a closer look at the checkbox `Render` until now. Seems like it relies on `Render` to be a `Semigroup` for manually combining a list of individual checkboxes into a single checkbox form (or further combine multiple checkbox forms into a single one) that can fit into the `( checkbox :: Render )` slot so the information specified in a unit `Render` is supposed to be local to a single checkbox which is somewhat in conflict with `columns` being a global property that we only want to specify once at the top level.

To make sense of it being a local property, I think we could use `Data.Ord.Max Int`. It could then be interpreted as choosing the widest among all the `columns` specified by the sub-forms. In the Halogen `render`, if `columns < 1` then we render the list of checkboxes vertically like before.

citizengabe replied:
I think we can let the `render` function decide to use `max` if we want that behavior.

boygao1992 replied:
The number of items in the `Array Int` depends on the number of times we use `Semigroup` instance to combine subforms which doesn't make much sense to me.

boygao1992 replied:
> I think we can let the `render` function decide to use `max` if we want that behavior.

The semantics should lie in the DSL layer not in the Halogen implementation.

citizengabe replied:
Well then I guess we need to decide what behavior we want. An argument could be made that `Data.Semigroup.First` is reasonable.

@davezuch do you have an opinion?

boygao1992 replied:
Either is fine with me as long as we document the behavior.

davezuch replied:
Wouldn't it make more sense to define a `RenderSet` type in this module with this `columns` property (do all `Render` types need to implement `Semigroup` or is it just a constraint of the checkbox `Render`?) and use that for `checkboxSet` instead of `Formlet.Render.Render`? https://github.com/citizennet/purescript-formlet-ocelot/blob/c09a5b746646241c33ca7da64ba21e6ca83f041e/src/Formlet/Ocelot/Checkbox.purs#L83C62-L83C62

citizengabe replied:
> Wouldn't it make more sense to define a `RenderSet` type in this module with this `columns` property (do all `Render` types need to implement `Semigroup` or is it just a constraint of the checkbox `Render`?) and use that for `checkboxSet` instead of `Formlet.Render.Render`? https://github.com/citizennet/purescript-formlet-ocelot/blob/c09a5b746646241c33ca7da64ba21e6ca83f041e/src/Formlet/Ocelot/Checkbox.purs#L83C62-L83C62

If we do this I think it makes sense to move `checkboxSet` to its own module. We're going to have to add another option for it to `Formlet.Ocelot.Render.Renders` anyway.

davezuch replied:
I'm fine with that.

boygao1992 replied:
> If we do this I think it makes sense to move `checkboxSet` to its own module. We're going to have to add another option for it to `Formlet.Ocelot.Render.Renders` anyway.

That sounds like re-engineering of the whole checkbox `Render` which might be an overkill for what we want to achieve.

citizengabe replied:
> > If we do this I think it makes sense to move `checkboxSet` to its own module. We're going to have to add another option for it to `Formlet.Ocelot.Render.Renders` anyway.
> 
> That sounds like re-engineering of the whole checkbox `Render` which might be an overkill for what we want to achieve.

`checkbox` and `checkboxSet` are only superficially similar. Their value and result types are different. In some ways, `checkboxSet` behaves more like a typeahead.

citizengabe replied:
I'm going to try splitting them apart and see where it takes me.

citizengabe committed:
Split out checkboxSet into its own module

Dave mentioned in https://github.com/citizennet/purescript-formlet-ocelot/pull/3#discussion_r1290502292
that `checkbox` and `checkboxSet` don't need to share the same `Render`
type and he's right. The two form components are only superficially
similar.

citizengabe committed:
Remove readonly from Render

This property is available on the config, so we don't need to track it
in the Render type.

davezuch commented on src/Formlet/Ocelot/CheckboxSet.purs:
```diff
@@ -0,0 +1,158 @@
+
+newtype Render a =
+  Render
+    { columns :: Int -- NOTE positive integer (> 0)
```
Maybe not worth the effort, but we could use a UInt for this: https://pursuit.purescript.org/packages/purescript-uint/7.0.0

citizengabe replied:
Since we're probably the only consumers of the library at the moment, I'd wait until either we keep making mistakes or someone else opens an issue.

davezuch reviewed (approved):
Changes look good! Though I didn't look closely at the ones that appear to just be moved code.

davezuch commented:
> It would be nice to add a storybook for these components though.

Yeah, I would advocate moving these libraries into the wildcat repo using `git subtree` if we anticipate working with them for a while longer (as opposed to sunsetting PS work in favor of the React rewrite), but I don't know how to judge that.

citizengabe commented:
Thanks for the reviews!
